File: command_pattern/python/2-bank-state-focus/bank_controller.py
from transaction import Transaction


class BankController:
    """This class will be our invoker. It will be responsible for executing the transactions."""

    def __init__(self) -> None:
        self.undo_stack = []
        self.redo_stack = []

    def execute(self, transaction: Transaction) -> None:
        transaction.execute()
        self.redo_stack.clear()
        self.undo_stack.append(transaction)

    def undo(self) -> None:
        if not self.undo_stack:
            return

        transaction = self.undo_stack.pop()
        transaction.undo()
        self.redo_stack.append(transaction)

    def redo(self) -> None:
        if not self.redo_stack:
            return

        transaction = self.redo_stack.pop()
        transaction.redo()
        self.undo_stack.append(transaction)

File: command_pattern/java/2-bank-state-focus/invoker/BankController.java
package invoker;

import java.util.ArrayList;
import java.util.List;

import command.Transaction;

public class BankController {
    
    private List<Transaction> undoStack;
    private List<Transaction> redoStack;

    public BankController() {
        this.undoStack = new ArrayList<>();
        this.redoStack = new ArrayList<>();
    }

    public void execute(Transaction transaction) {
        transaction.execute();
        undoStack.add(transaction);
        redoStack.clear();
    }

    public void undo() {
        if (undoStack.isEmpty()) {
            return;
        }
        Transaction transaction = undoStack.remove(undoStack.size() - 1);
        transaction.undo();
        redoStack.add(transaction);
    }

    public void redo() {
        if (redoStack.isEmpty()) {
            return;
        }
        Transaction transaction = redoStack.remove(redoStack.size() - 1);
        transaction.redo();
        undoStack.add(transaction);
    }
}
